homeassistant/components/notion/__init__.py

```diff
@@ -0,0 +1,272 @@
+from homeassistant.helpers.event import async_track_time_interval
+
+from .config_flow import configured_instances
+from .const import (
+    DATA_CLIENT, DEFAULT_SCAN_INTERVAL, DOMAIN, TOPIC_DATA_UPDATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_BRIDGE_MODE = 'bridge_mode'
+ATTR_BRIDGE_NAME = 'bridge_name'
+ATTR_SYSTEM_MODE = 'system_mode'
+ATTR_SYSTEM_NAME = 'system_name'
+
+DATA_LISTENER = 'listener'
+
+DEFAULT_ATTRIBUTION = 'Data provided by Notion'
+
+SENSOR_BATTERY = 'low_battery'
+SENSOR_DOOR = 'door'
+SENSOR_GARAGE_DOOR = 'garage_door'
+SENSOR_LEAK = 'leak'
+SENSOR_MISSING = 'missing'
+SENSOR_SAFE = 'safe'
+SENSOR_SLIDING = 'sliding'
+SENSOR_SMOKE_CO = 'alarm'
+SENSOR_TEMPERATURE = 'temperature'
+SENSOR_WINDOW_HINGED_HORIZONTAL = 'window_hinged_horizontal'
+SENSOR_WINDOW_HINGED_VERTICAL = 'window_hinged_vertical'
+
+BINARY_SENSOR_TYPES = {
+    SENSOR_BATTERY: ('Low Battery', 'battery'),
+    SENSOR_DOOR: ('Door', 'door'),
+    SENSOR_GARAGE_DOOR: ('Garage Door', 'garage_door'),
+    SENSOR_LEAK: ('Leak Detector', 'moisture'),
+    SENSOR_MISSING: ('Missing', 'connectivity'),
+    SENSOR_SAFE: ('Safe', 'door'),
+    SENSOR_SLIDING: ('Sliding Door/Window', 'door'),
+    SENSOR_SMOKE_CO: ('Smoke/Carbon Monoxide Detector', 'smoke'),
+    SENSOR_WINDOW_HINGED_HORIZONTAL: ('Hinged Window', 'window'),
+    SENSOR_WINDOW_HINGED_VERTICAL: ('Hinged Window', 'window'),
+}
+SENSOR_TYPES = {
+    SENSOR_TEMPERATURE: ('Temperature', 'temperature', '°C'),
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
+            cv.time_period,
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the Notion component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+    hass.data[DOMAIN][DATA_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    if conf[CONF_USERNAME] in configured_instances(hass):
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={'source': SOURCE_IMPORT},
+            data={
+                CONF_USERNAME: conf[CONF_USERNAME],
+                CONF_PASSWORD: conf[CONF_PASSWORD]
+            }))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up Notion as a config entry."""
+    from aionotion import async_get_client
+    from aionotion.errors import NotionError
+
+    session = aiohttp_client.async_get_clientsession(hass)
+
+    try:
+        client = await async_get_client(
+            config_entry.data[CONF_USERNAME],
+            config_entry.data[CONF_PASSWORD],
+            session)
+    except NotionError as err:
+        _LOGGER.error('Config entry failed: %s', err)
+        raise ConfigEntryNotReady
```

Comment:
Can you differentiate between authentication and connection errors ? Authentication errors are non-recoverable and would probably require an update from the user. Connection errors are recoverable. 

(this can be in a future PR)

Reply:
Great point. I'll need to update `aionotion` with a new exception, so I'll take your advice and do this as a separate PR.